src/posix/platform/alarm.c

```diff
@@ -103,53 +105,51 @@ void otPlatAlarmMicroStop(otInstance *aInstance)
     (void)aInstance;
     sIsUsRunning = false;
 }
+#endif // OPENTHREAD_CONFIG_ENABLE_PLATFORM_USEC_TIMER
 
 void platformAlarmUpdateTimeout(struct timeval *aTimeout)
 {
-    int32_t usRemaining = INT32_MAX;
-    int32_t msRemaining = INT32_MAX;
+    int64_t  remaining = INT32_MAX;
+    uint64_t now       = platformGetNow();
 
-    if (aTimeout == NULL)
-    {
-        return;
-    }
+    assert(aTimeout != NULL);
 
-    if (sIsUsRunning)
+    if (sIsMsRunning)
     {
-        usRemaining = (int64_t)(sUsAlarm - otPlatAlarmMicroGetNow());
+        remaining = (int32_t)(sMsAlarm - (uint32_t)(now / US_PER_MS));
```

Comment:
Is the following simpler?
```c
remaining = sMsAlarm * US_PER_MS - now;
```

Reply:
It won't be correct due to 32-bit wrapping of ms alarm time.

- Assume `uint64_t` value from `platformGetNow` is `(2^32 + 1) * 1000` usec (which adds up to ~50 days).
- `otPlatAlarmMilliGetNow` would return `(uint32_t)(platformGetNow() / US_PER_MS)` which is `1`
- Suppose we start a ms timer with 10ms interval at this time
- `sMsAlarm` uses `otPlatAlarmMilliGetNow()` so it will set `sMsAlarm <- 11`
- At this point `sMsAlarm * US_PER_MS - now` would give incorrect result.